include/deal.II/optimization/solver_bfgs.h

```diff
@@ -0,0 +1,402 @@
+  return line_search_signal.connect(slot);
+}
+
+
+
+template <class VectorType>
+boost::signals2::connection
+SolverBFGS<VectorType>::connect_preconditioner_slot(
+  const std::function<void(VectorType &                         g,
+                           const FiniteSizeHistory<VectorType> &s,
+                           const FiniteSizeHistory<VectorType> &y)> &slot)
+{
+  Assert(preconditioner_signal.empty(),
+         ExcMessage(
+           "One should not attach more than one preconditioner signal."));
+  return preconditioner_signal.connect(slot);
+}
+
+
+
+template <typename VectorType>
+void
+SolverBFGS<VectorType>::solve(
+  const std::function<typename VectorType::value_type(VectorType &x,
+                                                      VectorType &f)> &compute,
+  VectorType &                                                         x)
+{
+  // Also see scipy Fortran implementation
+  // https://github.com/scipy/scipy/blob/master/scipy/optimize/lbfgsb_src/lbfgsb.f
+  // and Octave-optim implementation:
+  // https://sourceforge.net/p/octave/optim/ci/default/tree/src/__bfgsmin.cc
+  LogStream::Prefix prefix("BFGS");
+
+  // default line search:
+  bool       first_step = true;
+  Number     f_prev     = 0.;
+  VectorType x0;
+  const auto default_line_min =
+    [&](Number &f, VectorType &x, VectorType &g, const VectorType &p) {
+      const Number f0 = f;
+      const Number g0 = g * p;
+      Assert(g0 < 0,
+             ExcMessage(
+               "Function does not decrease along the current direction"));
+
+      // save current solution value (to be used in line_search):
+      x0 = x;
+
+      // see scipy implementation
+      // https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.line_search.html#scipy.optimize.line_search
+      // and Eq. 2.6.8 in Fletcher 2013, Practical methods of optimization
+      Number df = f_prev - f;
+      Assert(first_step || df >= 0.,
+             ExcMessage("Function value is not decreasing"));
+      df = std::max(df, 100. * std::numeric_limits<Number>::epsilon());
+      // guess a reasonable first step:
+      const Number a1 = (first_step ? 1. : std::min(1., -1.01 * 2. * df / g0));
+      Assert(a1 > 0., ExcInternalError());
+      f_prev = f;
+
+      // 1D line-search function
+      const auto line_func =
+        [&](const Number &x_line) -> std::pair<Number, Number> {
+        x = x0;
+        x.add(x_line, p);
+        f                   = compute(x, g);
+        const Number g_line = g * p;
+        return std::make_pair(f, g_line);
+      };
+
+      // loose line search:
+      const auto res = LineMinimization::line_search<Number>(
+        line_func, f0, g0, LineMinimization::poly_fit<Number>, a1, 0.9, 0.001);
+
+      if (first_step)
+        first_step = false;
+
+      return res.first;
+    };
+
+  // provide default line search if no signal was attached
+  if (line_search_signal.empty())
+    {
+      x0.reinit(x);
```

Comment:
Why does this initialization only happen if the line search signal is empty?

Reply:
because it's not used otherwise.

Reply:
in particular, in the default line search it does
```
x = x0;
x.add(x_line, p);
```